service/application/src/main/java/com/solace/maas/ep/event/management/agent/subscriber/SolacePersistentMessageHandler.java

```diff
@@ -0,0 +1,97 @@
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Service
+@ConditionalOnExpression("${event-portal.gateway.messaging.standalone:false}== false && ${event-portal.managed:false} == true")
+public class SolacePersistentMessageHandler extends BaseSolaceMessageHandler implements MessageReceiver.MessageHandler {
+    private final Map<String, Class> cachedJSONDecoders = new HashMap();
+
+    private final Map<Class, MessageProcessor> messageProcessorsByClassType;
+    private final MessagingService messagingService;
+    private final EventPortalProperties eventPortalProperties;
+    private final PersistentMessageReceiver persistentMessageReceiver;
+
+    protected SolacePersistentMessageHandler(MessagingService messagingService,
+                                             EventPortalProperties eventPortalProperties,
+                                             List<MessageProcessor> messageProcessorList) {
+
+        super();
+        this.messagingService = messagingService;
+        this.eventPortalProperties = eventPortalProperties;
+        messageProcessorsByClassType = messageProcessorList.stream()
+                .collect(Collectors.toMap(MessageProcessor::supportedClass, Function.identity()));
+        Queue queue = determineQueue();
+        persistentMessageReceiver = buildPersistentMessageReceiver(queue);
+        persistentMessageReceiver.receiveAsync(this);
+        log.info("Binding to queue with persistent message handler");
+    }
+
+
+    @Override
+    public void onMessage(InboundMessage inboundMessage) {
+        String mopMessageSubclass = "";
+
+        try {
+            mopMessageSubclass = inboundMessage.getProperty(MOPConstants.MOP_MSG_META_DECODER);
+            String messageAsString = inboundMessage.getPayloadAsString();
+            Class messageClass = cachedJSONDecoders.get(mopMessageSubclass);
+            if (messageClass == null) {
+                messageClass = Class.forName(mopMessageSubclass);
+                cachedJSONDecoders.put(mopMessageSubclass, messageClass);
+            }
+            MessageProcessor processor = messageProcessorsByClassType.get(messageClass);
+            if (processor == null) {
+                throw new UnsupportedOperationException("Could not find message processor for message of class " + messageClass.getCanonicalName());
+            }
+            setupMDC(messageAsString, messageClass.getSimpleName());
+            log.trace("onMessage: {}\n{}", messageClass, messageAsString);
+            processor.processMessage(processor.castToMessageClass(toMessage(messageAsString, messageClass)));
+
+        } catch (Exception e) {
+            log.error("Error while processing inbound message from queue for mopMessageSubclass: {}", mopMessageSubclass);
+            throw new IllegalArgumentException(e);
+        } finally {
+            persistentMessageReceiver.ack(inboundMessage);
```

Comment:
We will ack the message even if there is any kind of exception. These is needed so that we don't create a poison pill scenario